File: files/300-399/346/CH-346.R
library(tidyverse)
library(readxl)
library(charcuterie)

path <- "300-399/346/CH-346 Filter.xlsx"
input <- read_excel(path, range = "B3:B10")
test <- read_excel(path, range = "F3:F6")

result = input %>%
  mutate(ID = map(ID, chars)) %>%
  mutate(last_char = map_chr(ID, ~ tail(.x, 1)),
         char_count = map(ID, ~ as.data.frame(table(.x)))) %>%
  unnest(char_count) %>%
  mutate(Freq = ifelse(.x == last_char, 0, Freq)) %>%
  filter(Freq >= 3) %>%
  mutate(ID = map_chr(ID, paste, collapse = "")) %>%
  select(ID)


all.equal(result, test, check.attributes = FALSE)
# [1] TRUE

File: files/300-399/346/CH-346.py
import pandas as pd
import numpy as np
from collections import Counter

path = "300-399/346/CH-346 Filter.xlsx"
input = pd.read_excel(path, usecols="B", skiprows=2, nrows=8)
test = pd.read_excel(path, usecols="F", skiprows=2, nrows=3).rename(columns=lambda col: col.replace('.1', ''))

def process_id(id_str):
    chars = list(str(id_str))
    last_char = chars[-1]
    char_count = Counter(chars)
    char_count[last_char] = 0
    filtered = {k: v for k, v in char_count.items() if v >= 3}
    return ''.join(chars) if filtered else None

input['processed'] = input['ID'].apply(process_id)
result = input.dropna(subset=['processed'])[['processed']]
result.columns = ['ID']

print(result.reset_index(drop=True).equals(test.reset_index(drop=True)))
